enterprise/backend/src/metabase_enterprise/enhancements/integrations/ldap.clj

```diff
@@ -40,14 +40,17 @@
     (apply dissoc m :objectclass (map (comp keyword u/lower-case-en) (ldap-sync-user-attributes-blacklist)))))
 
 (defn- attribute-synced-user
-  [{:keys [attributes email]}]
-  (when-let [user (db/select-one [User :id :last_login :login_attributes] :%lower.email (u/lower-case-en email))]
-    (let [syncable-attributes (syncable-user-attributes attributes)]
-      ;; Update User's `login_attributes` if needed
-      (if (and (not= (:login_attributes user) syncable-attributes)
-               (db/update! User (:id user) :login_attributes syncable-attributes))
-        (db/select-one [User :id :last_login] :id (:id user)) ; Reload updated user
-        user))))
+  [{:keys [attributes first-name last-name email]}]
+  (when-let [user (db/select-one [User :id :last_login :first_name :last_name :login_attributes]
+                                 :%lower.email (u/lower-case-en email))]
+            (let [syncable-attributes (syncable-user-attributes attributes)]
+              (do
+                (and (not= (:login_attributes user) syncable-attributes)
+                     (db/update! User (:id user) :login_attributes syncable-attributes))
+                (user/update-user! (:id user)
+                                   (default-impl/updated-name-part first-name (:first_name user))
+                                   (default-impl/updated-name-part last-name (:last_name user)))
```

Comment:
You can just call `db/update!` with a map. So merge just create a map with only the keys getting updated with `merge` or `cond->` and `assoc` or `clojure.data./diff` e.g.

```clj
(let [user-changes (merge
                    (when-not (= old-attributes new-attributes)
                      {:login_attributes new-attributes})
                    (when-not (= old-first-name new-first-name)
                      {:first_name new-first-name})
                    ...)]
  (db/update! User (u/the-id user) user-changes))
```